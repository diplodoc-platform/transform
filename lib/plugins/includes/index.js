const {bold} = require('chalk');

const {getFileTokens, getFullIncludePath} = require('../../utilsFS');
const {findBlockTokens} = require('../../utils');

const INCLUDE_REGEXP = /^{%\s*include\s*(notitle)?\s*\[(.+?)]\((.+?)\)\s*%}$/;

function stripTitleTokens(tokens) {
    if (tokens[0].type === 'heading_open' && tokens[2].type === 'heading_close') {
        tokens.splice(0, 3);
    }
}

function unfoldIncludes(state, path, options) {
    const {root, notFoundCb, log} = options;
    const {tokens} = state;
    let i = 0;

    while (i < tokens.length) {
        const openToken = tokens[i];
        const contentToken = tokens[i + 1];
        const closeToken = tokens[i + 2];
        let match;

        if (
            openToken.type === 'paragraph_open' &&
            contentToken.type === 'inline' &&
            (match = contentToken.content.match(INCLUDE_REGEXP)) &&
            closeToken.type === 'paragraph_close'
        ) {
            try {
                const [, keyword, /* description */, includePath] = match;
                const fullIncludePath = getFullIncludePath(includePath, root, path);
                const [pathname, hash] = fullIncludePath.split('#');

                if (!pathname.startsWith(root)) {
                    i++;

                    continue;
                }

                const fileTokens = getFileTokens(pathname, state, options);

                let includedTokens;
                if (hash) {
                    includedTokens = findBlockTokens(fileTokens, hash);
                } else {
                    includedTokens = fileTokens;
                }

                if (keyword === 'notitle') {
                    stripTitleTokens(includedTokens);
                }

                tokens.splice(i, 3, ...includedTokens);

                i += includedTokens.length;
            } catch (e) {
                if (notFoundCb) {
                    notFoundCb(e.path.replace(root, ''));
                }
                log.error(`Skip error: ${e}`);

                i++;
            }
        } else {
            i++;
        }
    }
}

function index(md, options) {
    const {path: optPath, log} = options;

    const plugin = (state) => {
        const {env} = state;
        const path = env.path || optPath;

        env.includes = env.includes || [];

        if (env.includes.includes(path)) {
            log.error(`Circular includes: ${bold(env.includes.concat(path).join(' â–¶ '))}`);
            process.exit(1);
        }

        env.includes.push(path);
        unfoldIncludes(state, path, options);
        env.includes.pop();
    };

    try {
        md.core.ruler.before('curly_attributes', 'includes', plugin);
    } catch (e) {
        md.core.ruler.push('includes', plugin);
    }
}

module.exports = index;
