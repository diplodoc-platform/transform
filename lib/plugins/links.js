const path = require('path');
const url = require('url');
const MarkdownIt = require('markdown-it');
const {
    isLocalUrl,
    isFileExists,
    getFileTokens,
    findBlockTokens,
    headingInfo,
    getSinglePageAnchorId,
} = require('../utils');
const {
    logEmptyLinkHref,
    logTitleRefLinkNotFound,
    logUnreachableLink,
} = require('../lintRules/pluginsRules');

const PAGE_LINK_REGEXP = /\.(md|ya?ml)$/i;

function defaultTransformLink(href) {
    const parsed = url.parse(href);

    return url.format({
        ...parsed,
        pathname: parsed.pathname.replace(PAGE_LINK_REGEXP, '.html'),
    });
}

function getTitleFromTokens(tokens) {
    let title = '';

    let i = 0;
    while (i < tokens.length) {
        const token = tokens[i];

        if (token.type === 'heading_open') {
            ({title = ''} = headingInfo(tokens, i));

            break;
        }

        i += 2;
    }

    return title;
}

const addTitle = ({hash, file, state, opts, isEmptyLink, tokens, idx, nextToken, href, currentPath}) => {
    const {log, lintOptions, disableLint} = opts;

    const id = hash && hash.slice(1);
    const fileTokens = getFileTokens(file, state, opts);
    const sourceTokens = id ? findBlockTokens(fileTokens, id) : fileTokens;
    const title = getTitleFromTokens(sourceTokens);

    if (title) {
        let textToken;

        if (isEmptyLink) {
            textToken = new state.Token('text', '', 0);
            tokens.splice(idx + 1, 0, textToken);
        } else {
            textToken = nextToken;
        }

        textToken.content = title;
    } else {
        logTitleRefLinkNotFound({
            lintOptions,
            commonOptions: {log, path: currentPath, href, disableLint},
        });
    }
};

// eslint-disable-next-line complexity
function processLink(state, tokens, idx, opts) {
    const {path: startPath, root, transformLink, notFoundCb, needSkipLinkFn, log, lintOptions, disableLint} = opts;
    const currentPath = state.env.path || startPath;
    const linkToken = tokens[idx];
    const nextToken = tokens[idx + 1];
    let href = linkToken.attrGet('href');

    if (!href) {
        logEmptyLinkHref({
            lintOptions,
            commonOptions: {log, path: startPath, disableLint},
        });
        return;
    }

    const {pathname, hash} = url.parse(href);
    let file;
    let fileExists;
    let isPageFile;

    if (!isLocalUrl(href)) {
        linkToken.attrSet('target', '_blank');
        linkToken.attrSet('rel', 'noreferrer noopener');
        return;
    }

    if (pathname) {
        file = path.resolve(path.parse(currentPath).dir, pathname);
        fileExists = isFileExists(file);
        isPageFile = PAGE_LINK_REGEXP.test(pathname);

        if (isPageFile && !fileExists) {
            let needShowError = true;
            if (needSkipLinkFn) {
                needShowError = !needSkipLinkFn(href);
            }

            if (notFoundCb && needShowError) {
                notFoundCb(file.replace(root, ''));
            }

            if (needShowError) {
                logUnreachableLink({
                    lintOptions,
                    commonOptions: {log, path: currentPath, href, disableLint},
                });
            }
        }
    } else if (hash) {
        file = startPath;
        fileExists = true;
        isPageFile = true;
    } else {
        return;
    }

    const isEmptyLink = nextToken.type === 'link_close';
    const isTitleRefLink = nextToken.type === 'text' && nextToken.content === '{#T}';
    if ((isEmptyLink || isTitleRefLink) && fileExists && isPageFile) {
        addTitle({hash, file, state, opts, isEmptyLink, tokens, idx, nextToken, href, currentPath});
    }

    let newPathname;
    if (currentPath !== startPath) {
        newPathname = path.relative(path.parse(startPath).dir, file);

        href = url.format({
            ...url.parse(href),
            pathname: newPathname,
        });
    }

    if (pathname || newPathname) {
        const transformer = transformLink || defaultTransformLink;
        linkToken.attrSet('href', transformer(href));
    }
}

function links(md, opts) {
    const plugin = (state) => {
        const tokens = state.tokens;
        let i = 0;

        state.md.disable(['includes', 'links'], true);

        while (i < tokens.length) {
            if (tokens[i].type === 'inline') {
                const childrenTokens = tokens[i].children;
                let j = 0;

                while (j < childrenTokens.length) {
                    const isLinkOpenToken = childrenTokens[j].type === 'link_open';
                    const tokenClass = childrenTokens[j].attrGet('class');

                    /*  Don't process anchor links */
                    const isYfmAnchor = tokenClass
                        ? tokenClass.includes('yfm-anchor')
                        : false;

                    if (isLinkOpenToken && !isYfmAnchor) {
                        processLink(state, childrenTokens, j, opts);
                    }

                    j++;
                }
            }

            i++;
        }

        state.md.enable(['includes', 'links'], true);
    };

    try {
        md.core.ruler.before('includes', 'links', plugin);
    } catch (e) {
        md.core.ruler.push('links', plugin);
    }
}

const replaceLinkHref = (input, href, newHref) => {
    /* Try not replace include syntax */
    return input.replace(`](${href})`, `](${newHref})`);
};

/* Replace the links to the markdown and yaml files if the singlePage option is passed in the options
*  Example: replace [Text](../../path/to/file.md#anchor) with [Text](#_path_to_file_anchor)
* */
links.collect = (input, options) => {
    const {root, path: startPath, singlePage} = options;

    if (!singlePage) {
        return;
    }

    let result = input;

    /* Syntax "{% include [Text](_includes/file.md) %}" is parsed as link. Need to ignore errors */
    const needSkipLinkFn = (href) => href.includes('_includes/');
    const transformLink = (href) => href;
    const md = new MarkdownIt()
        .use(links, {...options, transformLink, needSkipLinkFn});
    const tokens = md.parse(result, {});

    let i = 0;
    while (i < tokens.length) {
        if (tokens[i].type === 'inline') {
            const childrenTokens = tokens[i].children;
            let j = 0;

            while (j < childrenTokens.length) {
                const isLinkOpenToken = childrenTokens[j].type === 'link_open';

                if (!isLinkOpenToken) {
                    j++;
                    continue;
                }

                const linkToken = childrenTokens[j];
                const href = linkToken.attrGet('href');

                if (!href || !isLocalUrl(href) || href.includes('_includes/')) {
                    j++;
                    continue;
                }

                const {pathname, hash} = url.parse(href);
                if (pathname) {
                    const isPageFile = PAGE_LINK_REGEXP.test(pathname);
                    if (isPageFile) {
                        const newHref = getSinglePageAnchorId({root, currentPath: startPath, pathname, hash});
                        result = replaceLinkHref(result, href, newHref);
                    }
                } else if (hash) {
                    const newHref = getSinglePageAnchorId({root, currentPath: startPath, hash});
                    result = replaceLinkHref(result, href, newHref);
                }

                j++;
            }
        }

        i++;
    }

    // eslint-disable-next-line consistent-return
    return result;
};


module.exports = links;
