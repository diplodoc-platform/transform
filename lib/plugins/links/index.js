const path = require('path');
const url = require('url');
const {bold} = require('chalk');
const {
    isLocalUrl,
    findBlockTokens,
    headingInfo,
} = require('../../utils');
const {
    isFileExists,
    getFileTokens,
} = require('../../utilsFS');
const {PAGE_LINK_REGEXP} = require('./constants');

function defaultTransformLink(href) {
    const parsed = url.parse(href);

    return url.format({
        ...parsed,
        pathname: parsed.pathname.replace(PAGE_LINK_REGEXP, '.html'),
    });
}

function getTitleFromTokens(tokens) {
    let title = '';

    let i = 0;
    while (i < tokens.length) {
        const token = tokens[i];

        if (token.type === 'heading_open') {
            ({title = ''} = headingInfo(tokens, i));

            break;
        }

        i += 2;
    }

    return title;
}

const addTitle = ({hash, file, state, opts, isEmptyLink, tokens, idx, nextToken, href, currentPath, log}) => {
    const id = hash && hash.slice(1);
    const fileTokens = getFileTokens(file, state, opts);
    const sourceTokens = id ? findBlockTokens(fileTokens, id) : fileTokens;
    const title = getTitleFromTokens(sourceTokens);

    if (title) {
        let textToken;

        if (isEmptyLink) {
            textToken = new state.Token('text', '', 0);
            tokens.splice(idx + 1, 0, textToken);
        } else {
            textToken = nextToken;
        }

        textToken.content = title;
    } else {
        log.warn(`Title not found: ${bold(href)} in ${bold(currentPath)}`);
    }
};

// eslint-disable-next-line complexity
function processLink(state, tokens, idx, opts) {
    const {path: startPath, root, transformLink, notFoundCb, needSkipLinkFn, log} = opts;
    const currentPath = state.env.path || startPath;
    const linkToken = tokens[idx];
    const nextToken = tokens[idx + 1];
    let href = linkToken.attrGet('href');

    if (!href) {
        log.error(`Empty link in ${bold(startPath)}`);
        return;
    }

    const {pathname, hash} = url.parse(href);
    let file;
    let fileExists;
    let isPageFile;

    if (!isLocalUrl(href)) {
        linkToken.attrSet('target', '_blank');
        linkToken.attrSet('rel', 'noreferrer noopener');
        return;
    }

    if (pathname) {
        file = path.resolve(path.parse(currentPath).dir, pathname);
        fileExists = isFileExists(file);
        isPageFile = PAGE_LINK_REGEXP.test(pathname);

        if (isPageFile && !fileExists) {
            let needShowError = true;
            if (needSkipLinkFn) {
                needShowError = !needSkipLinkFn(href);
            }

            if (notFoundCb && needShowError) {
                notFoundCb(file.replace(root, ''));
            }

            if (needShowError) {
                log.error(`Link is unreachable: ${bold(href)} in ${bold(currentPath)}`);
            }
        }
    } else if (hash) {
        file = startPath;
        fileExists = true;
        isPageFile = true;
    } else {
        return;
    }

    const isEmptyLink = nextToken.type === 'link_close';
    const isTitleRefLink = nextToken.type === 'text' && nextToken.content === '{#T}';
    if ((isEmptyLink || isTitleRefLink) && fileExists && isPageFile) {
        addTitle({hash, file, state, opts, isEmptyLink, tokens, idx, nextToken, href, currentPath, log});
    }

    let newPathname;
    if (currentPath !== startPath) {
        newPathname = path.relative(path.parse(startPath).dir, file);

        href = url.format({
            ...url.parse(href),
            pathname: newPathname,
        });
    }

    if (pathname || newPathname) {
        const transformer = transformLink || defaultTransformLink;
        linkToken.attrSet('href', transformer(href));
    }
}

function index(md, opts) {
    const plugin = (state) => {
        const tokens = state.tokens;
        let i = 0;

        state.md.disable(['includes', 'links'], true);

        while (i < tokens.length) {
            if (tokens[i].type === 'inline') {
                const childrenTokens = tokens[i].children;
                let j = 0;

                while (j < childrenTokens.length) {
                    const isLinkOpenToken = childrenTokens[j].type === 'link_open';
                    const tokenClass = childrenTokens[j].attrGet('class');

                    /*  Don't process anchor links */
                    const isYfmAnchor = tokenClass
                        ? tokenClass.includes('yfm-anchor')
                        : false;

                    if (isLinkOpenToken && !isYfmAnchor) {
                        processLink(state, childrenTokens, j, opts);
                    }

                    j++;
                }
            }

            i++;
        }

        state.md.enable(['includes', 'links'], true);
    };

    try {
        md.core.ruler.before('includes', 'links', plugin);
    } catch (e) {
        md.core.ruler.push('links', plugin);
    }
}

module.exports = index;
