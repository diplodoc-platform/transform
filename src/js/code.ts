import {getEventTarget, isCustom} from './utils';

const BUTTON_SELECTOR = '.yfm-clipboard-button';

function copyToClipboard(text: string) {
    if (!text) {
        return Promise.resolve();
    }

    if (navigator.clipboard && typeof navigator.clipboard.writeText) {
        return navigator.clipboard.writeText(text);
    }
    const textarea = document.createElement('textarea');
    textarea.setAttribute('style', 'position: absolute; left: 1000%');
    textarea.textContent = text;
    document.body.append(textarea);

    textarea.select();
    document.execCommand('copy');

    document.body.removeChild(textarea);

    return Promise.resolve();
}

function notifySuccess(svgButton: HTMLElement | null) {
    if (!svgButton) {
        return;
    }

    const id = svgButton.getAttribute('data-animation');
    // @ts-expect-error
    const icon = svgButton.getRootNode().getElementById(`visibileAnimation-${id}`);

    if (!icon) {
        return;
    }

    icon.beginElement();
}

if (typeof document !== 'undefined') {
    document.addEventListener('click', (event) => {
        const target = getEventTarget(event) as HTMLElement;

        if (isCustom(event) || !target.matches(BUTTON_SELECTOR)) {
            return;
        }

        const parent = target.parentNode;

        if (!parent) {
            return;
        }

        const code = parent.querySelector<HTMLElement>('pre code');
        if (!code) {
            return;
        }

        copyToClipboard(code.innerText).then(() => {
            notifySuccess(parent.querySelector('.yfm-clipboard-icon'));
        });
    });
}
