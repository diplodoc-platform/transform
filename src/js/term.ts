import {getEventTarget, isCustom} from './utils';

const Selector = {
    TITLE: '.yfm .yfm-term_title',
    CONTENT: '.yfm .yfm-term_dfn',
};

const ClassName = {
    OPEN: 'open',
};

function createDefinitionElement(element: HTMLElement) {
    const termKey = element.getAttribute('term-key');

    const definitionTemplate = document.getElementById(
        termKey + '_template',
    ) as HTMLTemplateElement;
    const definitionElement = definitionTemplate?.content.cloneNode(true).firstChild as HTMLElement;

    document.body.append(definitionElement);

    return definitionElement;
}

function setDefinitionId(definitionElement: HTMLElement, termElement: HTMLElement): void {
    if (!definitionElement.getAttribute('id')) {
        const termKey = termElement.getAttribute('term-key');
        definitionElement?.setAttribute('id', termKey + '_element');
    }

    const termId = termElement.getAttribute('id') || Math.random().toString(36).substr(2, 8);
    definitionElement?.setAttribute('term-id', termId);
}

function setDefinitionPosition(definitionElement: HTMLElement, termElement: HTMLElement): void {
    const {x: termX, y: termY} = termElement.getBoundingClientRect();

    const relativeX = Number(definitionElement.getAttribute('relativeX'));
    const relativeY = Number(definitionElement.getAttribute('relativeY'));

    if (relativeX === termX && relativeY === termY) {
        return;
    }

    definitionElement.setAttribute('relativeX', String(termX));
    definitionElement.setAttribute('relativeY', String(termY));

    const offsetTop = 25;

    definitionElement.style.top = Math.abs(termY) + window.scrollY + offsetTop + 'px';
    definitionElement.style.left = termX + window.scrollX + 'px';
}

if (typeof document !== 'undefined') {
    document.addEventListener('click', (event) => {
        const openDefinition = document.getElementsByClassName(
            'yfm yfm-term_dfn open',
        )[0] as HTMLElement;
        const target = getEventTarget(event) as HTMLElement;

        const termId = target.getAttribute('id');
        const termKey = target.getAttribute('term-key');
        let definitionElement = document.getElementById(termKey + '_element');

        if (termKey && !definitionElement) {
            definitionElement = createDefinitionElement(target);
        }

        const isSameTerm = openDefinition && termId === openDefinition.getAttribute('term-id');
        if (isSameTerm) {
            openDefinition.classList.remove(ClassName.OPEN);
            return;
        }

        const isTargetDefinitionContent = target.closest(
            [Selector.CONTENT.replace(' ', ''), ClassName.OPEN].join('.'),
        );

        if (openDefinition && !isTargetDefinitionContent) {
            openDefinition.classList.remove(ClassName.OPEN);
        }

        if (isCustom(event) || !target.matches(Selector.TITLE) || !definitionElement) {
            return;
        }

        setDefinitionId(definitionElement, target);
        setDefinitionPosition(definitionElement, target);

        definitionElement.classList.toggle(ClassName.OPEN);
    });

    document.addEventListener('keydown', (event) => {
        const openDefinition = document.getElementsByClassName(
            'yfm yfm-term_dfn open',
        )[0] as HTMLElement;
        if (event.key === 'Escape' && openDefinition) {
            openDefinition.classList.remove(ClassName.OPEN);
        }
    });

    window.addEventListener('resize', () => {
        const openDefinition = document.getElementsByClassName(
            'yfm yfm-term_dfn open',
        )[0] as HTMLElement;

        if (!openDefinition) {
            return;
        }

        const termId = openDefinition.getAttribute('term-id') || '';
        const termElement = document.getElementById(termId);

        if (!termElement) {
            return;
        }

        setDefinitionPosition(openDefinition, termElement);
    });
}
