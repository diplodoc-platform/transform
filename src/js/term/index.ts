import {
    Selector,
    openClass,
    openDefinitionClass,
    createDefinitionElement,
    setDefinitionId,
    setDefinitionPosition,
    closeDefinition,
} from './utils';
import {getEventTarget, isCustom} from '../utils';

if (typeof document !== 'undefined') {
    document.addEventListener('click', (event) => {
        const openDefinition = document.getElementsByClassName(
            openDefinitionClass,
        )[0] as HTMLElement;
        const target = getEventTarget(event) as HTMLElement;

        const termId = target.getAttribute('id');
        const termKey = target.getAttribute('term-key');
        let definitionElement = document.getElementById(termKey + '_element');

        if (termKey && !definitionElement) {
            definitionElement = createDefinitionElement(target);
        }

        const isSameTerm = openDefinition && termId === openDefinition.getAttribute('term-id');
        if (isSameTerm) {
            closeDefinition(openDefinition);
            return;
        }

        const isTargetDefinitionContent = target.closest(
            [Selector.CONTENT.replace(' ', ''), openClass].join('.'),
        );

        if (openDefinition && !isTargetDefinitionContent) {
            closeDefinition(openDefinition);
        }

        if (isCustom(event) || !target.matches(Selector.TITLE) || !definitionElement) {
            return;
        }

        setDefinitionId(definitionElement, target);
        setDefinitionPosition(definitionElement, target);

        definitionElement.classList.toggle(openClass);
    });

    document.addEventListener('keydown', (event) => {
        const openDefinition = document.getElementsByClassName(
            openDefinitionClass,
        )[0] as HTMLElement;
        if (event.key === 'Escape' && openDefinition) {
            closeDefinition(openDefinition);
        }
    });

    window.addEventListener('resize', () => {
        const openDefinition = document.getElementsByClassName(
            openDefinitionClass,
        )[0] as HTMLElement;

        if (!openDefinition) {
            return;
        }

        const termId = openDefinition.getAttribute('term-id') || '';
        const termElement = document.getElementById(termId);

        if (!termElement) {
            openDefinition.classList.toggle(openClass);
            return;
        }

        setDefinitionPosition(openDefinition, termElement);
    });
}
