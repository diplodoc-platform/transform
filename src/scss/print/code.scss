@media print {
    .yfm {
        .yfm-clipboard {
            & > pre > code {
                white-space: pre-wrap;
            }
        }
    }
}