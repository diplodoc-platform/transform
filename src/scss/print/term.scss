@media print {
    .yfm {
        .yfm-term {
            &_title {
                color: currentColor;
                border-bottom: none;
            }
        }
    }
}