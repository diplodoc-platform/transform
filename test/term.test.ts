import {dirname, resolve} from 'path';
import {readFileSync} from 'fs';
import transform from '../src/transform';
import links from '../src/transform/plugins/links';
import term from '../src/transform/plugins/term';
import includes from '../src/transform/plugins/includes';
import code from '../src/transform/plugins/code';

const mocksPath = require.resolve('./utils.ts');

const transformYfm = (text: string, path?: string) => {
    const {
        result: {html},
    } = transform(text, {
        plugins: [includes, links, code, term],
        path: path || mocksPath,
        root: dirname(path || mocksPath),
    });
    return html;
};

const clearRandomId = (str: string) => {
    const clearRandomId = new RegExp(/<([i\s]+).*?id="([^"]*?)".*?>(.+?)/, 'g');
    const randomId = clearRandomId.exec(str);

    return randomId ? str.replace(randomId[2], '') : str;
};

describe('Terms', () => {
    test('Should create term in text with definition template', () => {
        const inputPath = resolve(__dirname, './mocks/term/term.md');
        const input = readFileSync(inputPath, 'utf8');
        const result = transformYfm(input, inputPath);

        expect(clearRandomId(result)).toEqual(
            '<h1>Web</h1>\n' +
                '<p>The <i class="yfm yfm-term_title" term-key=":html" aria-describedby=":html_element" id="">HTML</i> specification</p>\n' +
                '<template id=":html_template"><dfn class="yfm yfm-term_dfn" id=":html_element" role="tooltip"><p>The HyperText Markup Language or <strong>HTML</strong> is the standard markup language for documents designed to be displayed in a web browser.</p>\n' +
                '</dfn></template>',
        );
    });

    test('Should create term in table with definition template', () => {
        const inputPath = resolve(__dirname, './mocks/term/table.md');
        const input = readFileSync(inputPath, 'utf8');
        const result = transformYfm(input, inputPath);

        expect(clearRandomId(result)).toEqual(
            '<h1>Web</h1>\n' +
                '<table>\n' +
                '<thead>\n' +
                '<tr>\n' +
                '<th>Language</th>\n' +
                '<th style="text-align:center">Initial release</th>\n' +
                '</tr>\n' +
                '</thead>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td><i class="yfm yfm-term_title" term-key=":html" aria-describedby=":html_element" id="">HTML</i></td>\n' +
                '<td style="text-align:center">1993</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n' +
                '<template id=":html_template"><dfn class="yfm yfm-term_dfn" id=":html_element" role="tooltip"><p>The HyperText Markup Language or <strong>HTML</strong> is the standard markup language for documents designed to be displayed in a web browser.</p>\n' +
                '</dfn></template>',
        );
    });

    test('Should create term in code with definition template', () => {
        const inputPath = resolve(__dirname, './mocks/term/code.md');
        const input = readFileSync(inputPath, 'utf8');
        const result = transformYfm(input, inputPath);

        expect(clearRandomId(result)).toEqual(
            '<h1>Web</h1>\n' +
                '\n' +
                '    <div class="yfm-clipboard">\n' +
                '    <pre><code class="hljs"><i class="yfm yfm-term_title" term-key=":html" id="">HTML</i>: Lorem\n' +
                '</code></pre>\n' +
                '\n' +
                '    <svg width="16" height="16" viewBox="0 0 24 24" class="yfm-clipboard-button" data-animation="3">\n' +
                '        <path\n' +
                '            fill="currentColor"\n' +
                '            d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"\n' +
                '        />\n' +
                '        <path\n' +
                '            stroke="currentColor"\n' +
                '            fill="transparent"\n' +
                '            strokeWidth="1.5"\n' +
                '            d="M9.5 13l3 3l5 -5"\n' +
                '            visibility="hidden"\n' +
                '        >\n' +
                '            <animate\n' +
                '                id="visibileAnimation-3"\n' +
                '                attributeName="visibility"\n' +
                '                from="hidden"\n' +
                '                to="visible"\n' +
                '                dur="0.2s"\n' +
                '                fill="freeze"\n' +
                '                begin=""\n' +
                '            />\n' +
                '            <animate\n' +
                '                id="hideAnimation-3"\n' +
                '                attributeName="visibility"\n' +
                '                from="visible"\n' +
                '                to="hidden"\n' +
                '                dur="1s"\n' +
                '                begin="visibileAnimation-3.end+1"\n' +
                '                fill="freeze"\n' +
                '            />\n' +
                '        </path>\n' +
                '    </svg>\n' +
                '    </div>\n' +
                '<template id=":html_template"><dfn class="yfm yfm-term_dfn" id=":html_element" role="tooltip"><p>The HyperText Markup Language or <strong>HTML</strong> is the standard markup language for documents designed to be displayed in a web browser.</p>\n' +
                '</dfn></template>',
        );
    });

    test('Term should use content from include', () => {
        const inputPath = resolve(__dirname, './mocks/term/includeContent.md');
        const input = readFileSync(inputPath, 'utf8');
        const result = transformYfm(input, inputPath);

        expect(clearRandomId(result)).toEqual(
            '<h1>Web</h1>\n' +
                '<p>The <i class="yfm yfm-term_title" term-key=":html" aria-describedby=":html_element" id="">HTML</i> specification</p>\n' +
                '<template id=":html_template"><dfn class="yfm yfm-term_dfn" id=":html_element" role="tooltip"><p>The HyperText Markup Language or <strong>HTML</strong> is the standard markup language for documents designed to be displayed in a web browser.</p>\n' +
                '</dfn></template>',
        );
    });
});
