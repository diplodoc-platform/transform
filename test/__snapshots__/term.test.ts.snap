// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terms Should create term in code with definition template 1`] = `
"<template id=\\":html_template\\" label=\\"html\\"><dfn class=\\"yfm yfm-term_dfn\\" id=\\":html_element\\" role=\\"tooltip\\"><p>The HyperText Markup Language or <strong>HTML</strong> is the standard markup language for documents designed to be displayed in a web browser.</p>
</dfn></template><h1>Web</h1>

    <div class=\\"yfm-clipboard\\">
        <pre><code class=\\"hljs\\"><i class=\\"yfm yfm-term_title\\" term-key=\\":html\\" id=\\"\\">HTML</i>: Lorem
</code></pre>

        <button class=\\"yfm-clipboard-button\\">
            <svg width=\\"16\\" height=\\"16\\" viewBox=\\"0 0 24 24\\" class=\\"yfm-clipboard-icon\\" data-animation=\\"10\\">
                <path fill=\\"currentColor\\" d=\\"M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z\\"></path>
                <path stroke=\\"currentColor\\" fill=\\"transparent\\" stroke-width=\\"1.5\\" d=\\"M9.5 13l3 3l5 -5\\" visibility=\\"hidden\\">
                    <animate id=\\"visibileAnimation-10\\" attributeName=\\"visibility\\" from=\\"hidden\\" to=\\"visible\\" dur=\\"0.2s\\" fill=\\"freeze\\" begin></animate>
                    <animate id=\\"hideAnimation-10\\" attributeName=\\"visibility\\" from=\\"visible\\" to=\\"hidden\\" dur=\\"1s\\" begin=\\"visibileAnimation-10.end+1\\" fill=\\"freeze\\"></animate>
                </path>
            </svg>
        </button>
    </div>
"
`;

exports[`Terms Should create term in table with definition template 1`] = `
"<template id=\\":html_template\\" label=\\"html\\"><dfn class=\\"yfm yfm-term_dfn\\" id=\\":html_element\\" role=\\"tooltip\\"><p>The HyperText Markup Language or <strong>HTML</strong> is the standard markup language for documents designed to be displayed in a web browser.</p>
</dfn></template><h1>Web</h1>
<table>
<thead>
<tr>
<th>Language</th>
<th style=\\"text-align:center\\">Initial release</th>
</tr>
</thead>
<tbody>
<tr>
<td><i class=\\"yfm yfm-term_title\\" term-key=\\":html\\" role=\\"term\\" aria-describedby=\\":html_element\\" tabindex=\\"0\\" id=\\"\\">HTML</i></td>
<td style=\\"text-align:center\\">1993</td>
</tr>
</tbody>
</table>
"
`;

exports[`Terms Should create term in text with definition template 1`] = `
"<template id=\\":html_template\\" label=\\"html\\"><dfn class=\\"yfm yfm-term_dfn\\" id=\\":html_element\\" role=\\"tooltip\\"><p>The HyperText Markup Language or <strong>HTML</strong> is the standard markup language for documents designed to be displayed in a web browser.</p>
</dfn></template><h1>Web</h1>
<p>The <i class=\\"yfm yfm-term_title\\" term-key=\\":html\\" role=\\"term\\" aria-describedby=\\":html_element\\" tabindex=\\"0\\" id=\\"\\">HTML</i> specification</p>
"
`;

exports[`Terms Term should use content from include 1`] = `
"<template id=\\":html_template\\" label=\\"html\\"><dfn class=\\"yfm yfm-term_dfn\\" id=\\":html_element\\" role=\\"tooltip\\"><p>The HyperText Markup Language or <strong>HTML</strong> is the standard markup language for documents designed to be displayed in a web browser.</p>
</dfn></template><h1>Web</h1>
<p>The <i class=\\"yfm yfm-term_title\\" term-key=\\":html\\" role=\\"term\\" aria-describedby=\\":html_element\\" tabindex=\\"0\\" id=\\"\\">HTML</i> specification</p>
"
`;
